### PR TITLE
drivers/ws281x: add RMT hardware support for ESP32x SoCs

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides support for ESP32x RMT that is used to generate WS2812B signals.

All ESP32x SoCs have a [Remote Control Peripheral (RMT)](https://docs.espressif.com/projects/esp-idf/en/v4.4.4/esp32/api-reference/peripherals/rmt.html) that can be used to generate digital waveforms, such as NEC remote control signals or WS2812B RGB LED signals. Each RMT peripheral has either 4 or 8 channels. Some ESP32x SoCs support configuring the clock sources used for each channel separately, while other ESP32x SoCs can only use a single clock source for all channels.

The advantages using the RMT are that the CPU is not busy when generating the WS218x signal and the phase times are accurate to nanoseconds. The disadvantage is that the configuration of the RMT is so complex due to its flexibility that the [ESP-IDF high-level driver for RMT](https://docs.espressif.com/projects/esp-idf/en/v4.4.4/esp32c3/api-reference/peripherals/rmt.html) is used.This requires about 6 kBytes more ROM, about 3 kBytes more IRAM, and 88 bytes more RAM.

Therefore, either the RMT hardware implementation (module `ws2812x_esp32_hw`) as well as the bit-banging software implementation  (module  `ws2812x_esp32_hw`) can be used. RMT hardware implementation (module `ws2812x_esp32_hw`) is used by default.

Timing with hardware implementation:
ws281x_esp32_rmt

### Testing procedure

`tests/driver_ws281x` should still work, for example:
```
CFLAGS='-DWS281X_PARAM_PIN=GPIO45 -DWS281X_PARAM_NUMOF=47' BOARD=esp32s3-devkit make -j8 -C tests/driver_ws281x flash
```
Test Output:
https://user-images.githubusercontent.com/31932013/227791753-e6d2924e-3364-4387-870f-a56d3a9a8a80.mp4

### CODE REVIEW

File: boards/common/esp32/include/periph_conf_common.h [hunk 1]
```diff
@@ -251,6 +251,39 @@ static const pwm_config_t pwm_config[] =
 
 /** @} */
 
+/**
+ * @name    RMT configuration
```
Comment (maribu) [inline boards/common/esp32/include/periph_conf_common.h:255]:
```suggestion
 * @name    RMT configuration
 * @{
```

File: boards/common/esp32c3/include/periph_conf_common.h [hunk 2]
```diff
@@ -233,6 +233,39 @@ static const pwm_config_t pwm_config[] =
 
 /** @} */
 
+/**
+ * @name    RMT configuration
+ */
+
+/**
+ * @brief   RMT channel configuration
```
Comment (maribu) [inline boards/common/esp32c3/include/periph_conf_common.h:241]:
```suggestion
 * @brief   RMT channel configuration
 * @{
```

File: boards/common/esp32s2/include/periph_conf_common.h [hunk 3]
```diff
@@ -252,6 +252,39 @@ static const pwm_config_t pwm_config[] =
 
 /** @} */
 
+/**
+ * @name    RMT configuration
```
Comment (maribu) [inline boards/common/esp32s2/include/periph_conf_common.h:256]:
```suggestion
 * @name    RMT configuration
 * @{
```

File: boards/common/esp32s3/include/periph_conf_common.h [hunk 4]
```diff
@@ -252,6 +252,39 @@ static const pwm_config_t pwm_config[] =
 
 /** @} */
 
+/**
+ * @name    RMT configuration
```
Comment (maribu) [inline boards/common/esp32s3/include/periph_conf_common.h:256]:
```suggestion
 * @name    RMT configuration
 * @{
```

File: cpu/esp32/esp-idf/include/driver/gpio.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#ifndef DRIVER_GPIO_H
```
Comment (maribu) [inline cpu/esp32/esp-idf/include/driver/gpio.h:1]: The CI is upset about missing copyright and `extern "C" ` boilerplates

File: cpu/esp32/include/periph_cpu.h [hunk 6]
```diff
@@ -548,6 +548,39 @@ typedef struct {
...
+ *
+ * Each RMT channel is mapped to a GPIO. The configured mappings are used
+ * by the drivers that use the RMT peripheral to determine the RMT channel
+ * for a given GPIO.
+ */
+typedef struct {
+    uint8_t channel;    /**< channel index */
+    gpio_t gpio;        /**< GPIO used as RMT channel */
+} rmt_channel_config_t;
+
```
Comment (maribu) [inline cpu/esp32/include/periph_cpu.h:573]:
```suggestion
```

(Two empty lines)

**Review by maribu — COMMENTED** [review]
lgtm. I'll take a second look tomorrow on a big monitor just in case. I tend to do betters on that compared to a review on a phone screen 😅

File: cpu/esp32/esp-idf/include/driver/gpio.h (hunk 5 above)
Comment (maribu) [reply cpu/esp32/esp-idf/include/driver/gpio.h:1]: Other than the missing boilerplates, this looks good to me. Please squash at will.

**Comment by gschorcht** [comment]
:roll_eyes: I had to add `CONFIG_MODULE_WS281X_ESP32_HW=n` to the CI boards to enable `ws281x_esp32_sw` also with `TEST_KCONFIG=1`.

**Comment by maribu** [comment]
bors retry

**Comment by maribu** [comment]
Ah, squashing is still missing

**Comment by gschorcht** [comment]
> Ah, squashing is still missing

Yes, by intention to show the last changes.

**Comment by maribu** [comment]
bors retry

**Pull request merged**